rfc-113-expanding-draft-access-for-unauthenticated-users.md

```diff
@@ -0,0 +1,212 @@
+- Publishing API will be changed to support the sharing of auth-bypass ids for an
+  edition through link expansion rules.
+- Publishing API and Content Store will have their access limit code decoupled
+  from auth-bypass ids. This will be done to remove a frequent source of
+  confusion that conflates the separate processes of access limiting and
+  auth-bypass.
+- Collections Publisher will be changed to include a flag in auth-bypass tokens
+  that provide draft asset access. To reflect the increased access of a token
+  they will be given an expiry time (anticipated to be 1 month). Collections
+  Publisher will also be updated to include origin information in the token
+  (such as creator id and content id) to allow a token to be audited. The
+  [auth-bypass token documentation][auth-bypass-token-docs] will be updated to
+  reflect this as the best practice for future auth-bypass tokens
+
+### Rejected alternative approaches
+
+There were a number of alternative ideas which were considered to provide fact
+checking for step-by-steps and were ultimately rejected. This is a quick
+compilation of them and the reasons why they were rejected.
+
+#### Providing a link to preview the embedded part of a step-by-step distinct from the content
+
+This would involve creating a page which shows just the step-by-step
+navigation component that is embedded on linked pages. This was deemed
+insufficient as fact checkers are often comparing the step-by-step navigation
+with the context of the specific guidance pages.
+
+#### Providing a means to embed the step-by-step component into live content
+
+This would involve a means of injecting the step-by-step navigation component
+into live GOV.UK pages via JavaScript. This would prevent a fact checker of a
+step-by-step from viewing any draft content except the step-by-step
+navigation itself. This was quickly rejected as it was considered to be a
+common scenario for the pages that are part of a step-by-step
+to also be drafts.
+
+#### Exclude asset access from fact-checking
+
+Assets are the area of access that provide the most complication since they
+are distinct from the Publishing API link system. An option was to exclude
+them from the access checks and have them return forbidden errors if a
+step-by-step page used them - this would typically be most noticeable on an
+embedded image.
+
+This option was rejected due to it leaving a problem that could be easily
+anticipated and likely having to fall to a support team to resolve or explain
+whenever it occurs.
+
+#### Providing a fact checker access to all resources on the draft stack that are not access limited
+
+An approach that was considered was to allow the fact checker to have access
+to all pages on the draft stack that are not access limited. This would grant
+them similar access to an authenticated user. This offered the advantage that
+it allows simplifying the logic involved in checking tokens and is consistent
+with privacy expectations.
```

Comment:
There's another approach similar to this that I can't see referenced in the RFC, so thought I'd raise just for the record. It would be:

* If step by step page has a valid token in its URL...
* ...then append tokens to each of the links to other pages

The advantage would be that we're not granting blanket access to the entire draft stack, only to the pages linked to from the step by step. The disadvantages are twofold: 1) it would complicate the renderer, and 2) we don't know what each linked page's token should be.

The latter would probably require a change to how tokens are generated. My understanding is that it is up to the application to create the token and to tell publishing-api what the token is, but if we made it the responsibility of publishing-api to generate its own tokens, we could retrieve the token for each page by querying its API. Admittedly this would mean any application on GOV.UK could query the token of content published by other applications, which is an area of concern.

Reply:
@ChrisBAshton Thanks for the suggestion - I know we chatted IRL so I'll just summarise a bit here.

> There's another approach similar to this that I can't see referenced in the RFC, so thought I'd raise just for the record. It would be:
> 
> * If step by step page has a valid token in its URL...
> * ...then append tokens to each of the links to other pages
> 
> The advantage would be that we're not granting blanket access to the entire draft stack, only to the pages linked to from the step by step. The disadvantages are twofold: 1) it would complicate the renderer, and 2) we don't know what each linked page's token should be.

The token itself doesn't provide blanket access to the draft stack - just specific pages and most draft assets. 

As I understand this idea seems to be more about the means to distribute the token rather than the access level. 

To do this we'd either need to adjust the rendering applications involved to know about the potential for tokens and then be able to append them to URLs, which could involve a lot of draft specific changes. This is problematic since we purposefully keep apps distinct from knowing whether they are not the draft stack or not as much as possible.

A thing we could do is change authenticating proxy to rewrite the HTML of pages that are returned so all links have the token on them, which would effectively blanket share the token to all links that matched a pattern (like an old school session id). Though this would greatly increase the code of authenticating proxy and carry with it a number of challenges resulting from the HTTP rewriting. Although on the plus side every URL would be a sharable.

> The latter would probably require a change to how tokens are generated. My understanding is that it is up to the application to create the token and to tell publishing-api what the token is, but if we made it the responsibility of publishing-api to generate its own tokens, we could retrieve the token for each page by querying its API. Admittedly this would mean any application on GOV.UK could query the token of content published by other applications, which is an area of concern.

Applications still need to know the auth_bypass_id so they can share the token with a user. So if Publishing API did create them Collections Publisher would need to know it to share it. 

Retrieving the token at the point of rendering the page isn't ideal for a few reasons (performance being a key one) but one thing we'd want to avoid is that by sharing different pages tokens that could be actually used to grant users access to content that is access limited. For example say I have document "B" that is access limited to just me and you have a step by step that links to "B". You shouldn't be able to get the auth_bypass_id I created for B since you don't have access.